### PR TITLE
Add ProcessOrphans (move-only)

### PR DESCRIPTION
**Pull request by maflcko**
Moving the orphan processing into a new function makes it possible to call it from other places. E.g. `sendrawtransaction` or when a transaction is received via a `NetMsgType` different from `TX`.

Also, add test coverage for the function in our ~unit~ functional test framework.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -2179,9 +2250,6 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             AcceptToMemoryPool(mempool, state, ptx, &fMissingInputs, &lRemovedTxn, false /* bypass_limits */, 0 /* nAbsurdFee */)) {
```
Comment (Empact) [inline src/net_processing.cpp]: Looks like `lRemovedTxn` still needs to be cleared in this method, unless moving it into `ProcessOrphans` is appropriate.

**Comment by Empact** [comment]
Concept ACK, both for the addition of testing and due to the fact that ProcessMessage is >1300 lines at the moment. Would be great if other applications of the method were ready, to help inform its interface.

**Comment by skeees** [comment]
This is a move-only commit - so maybe not the right place - but it might be nice to separate the misbehaving peer punishment logic from the re-attempting to accept orphans to the mempool logic - good first step though

File: src/net_processing.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/net_processing.cpp]: Sorry, that indeed looks ugly right now. I will rebase this on #11775, which removes `lRemovedTxn` in whole.

File: test/functional/p2p_invalid_tx.py [hunk 2]
```diff
@@ -44,12 +65,48 @@ def run_test(self):
...
+        self.log.info('Send the withhold tx ... ')
+        assert_equal(2, len(node.getpeerinfo()))
+        # The tx itself is successfully accepted, but we should get disconnected anyway
+        node.p2p.send_txs_and_test([tx_withhold], node, success=True, expect_disconnect=False)
+        assert_equal(1, len(node.getpeerinfo()))
 
-        # Verify valid transaction
-        tx1 = create_transaction(block1.vtx[0], 0, b'', 50 * COIN - 12000)
-        node.p2p.send_txs_and_test([tx1], node, success=True)
+        assert_equal({t.hash for t in [tx_withhold, tx_orphan_1, tx_orphan_2_valid]}, set(node.getrawmempool()))
```
Comment (ryanofsky) [inline test/functional/p2p_invalid_tx.py]: Could make sense to rebase the test commit before the refactoring commit, to make it clear existing code is supposed to pass the new test.

Or maybe just put the new test in a standalone PR if the refactoring is going to be held up by #11775 anyway (https://github.com/bitcoin/bitcoin/pull/12935#discussion_r180830562).

**Review by ryanofsky — COMMENTED** [review]
utACK aaaa4569b1ef49ce6f031325f03d171de55d8419 for both test code and moveonly refactoring code (I didn't have any comments on the moveonly code).

File: test/functional/p2p_invalid_tx.py [hunk 3]
```diff
@@ -6,24 +6,45 @@
...
-        self.extra_args = [["-whitelist=127.0.0.1"]]
+
+    def reconnect_p2p(self, additional_conn=False):
+        """Add a P2P connection to the node.
+
+        The node gets disconnected several times in this test. This helper
+        method reconnects the p2p and restarts the network thread."""
+
+        network_thread_join()
+        self.nodes[0].disconnect_p2ps()
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply test/functional/p2p_invalid_tx.py]: Agree. Should be fixed in both places.
